src/cargo/util/toml/mod.rs

```diff
@@ -1992,15 +2013,23 @@ fn detailed_dep_to_dependency<P: ResolveToPath + Clone>(
     }
 
     if let Some(p) = orig.public {
-        manifest_ctx
-            .features
-            .require(Feature::public_dependency())?;
+        let public_feature = manifest_ctx.features.require(Feature::public_dependency());
+        let with_z_public = manifest_ctx.gctx.cli_unstable().public_dependency;
+        let with_public_feature = public_feature.is_ok();
+        if !with_public_feature && (!with_z_public && !manifest_ctx.gctx.nightly_features_allowed) {
+            public_feature?;
+        }
 
         if dep.kind() != DepKind::Normal {
-            bail!("'public' specifier can only be used on regular dependencies, not {:?} dependencies", dep.kind());
+            match (with_public_feature, with_z_public) {
+                (true, _) => bail!("'public' specifier can only be used on regular dependencies, not {:?} dependencies", dep.kind()),
+                (_, true) => bail!("'public' specifier can only be used on regular dependencies, not {:?} dependencies", dep.kind()),
```

Comment:
nit: de-duplicate error messages

Reply:
It reminds me that I can do this：
`(true, _) | (_, true) => bail!("'public' specifier can only be used on regular dependencies, not {:?} dependencies", dep.kind()),`